### PR TITLE
[feature](nereids) Support query on specific partitions 

### PR DESCRIPTION
**Pull request by Kikyou1997**
# Proposed changes

Issue Number: close #xxx

## Problem summary

Describe your changes.

## Checklist(Required)

1. Does it affect the original behavior: 
    - [ ] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [ ] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [ ] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [ ] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 34.75 seconds
 load time: 629 seconds
 storage size: 17123253915 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221221131522_clickbench_pr_66370.html

File: fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java [hunk 1]
```diff
@@ -410,7 +412,18 @@ private LogicalPlan withCheckPolicy(LogicalPlan plan) {
...
         List<String> tableId = visitMultipartIdentifier(ctx.multipartIdentifier());
-        LogicalPlan checkedRelation = withCheckPolicy(new UnboundRelation(RelationUtil.newRelationId(), tableId));
+        List<String> partitionNames = new ArrayList<>();
+        if (ctx.specified_partition() != null) {
+            if (ctx.specified_partition().identifier() != null) {
+                partitionNames.add(ctx.specified_partition().identifier().getText());
+            } else {
+                partitionNames.addAll(Arrays
+                        .stream(ctx.specified_partition().identifierList().identifierSeq().getText().split(","))
+                        .collect(Collectors.toList()));
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/parser/LogicalPlanBuilder.java:422]:
```suggestion
                partitionNames.addAll(visitIdentifierList(ctx.specified_partition().identifierList()));
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java [hunk 2]
```diff
@@ -52,19 +57,20 @@ public Rule build() {
...
+                    return bindWithDbNameFromNamePart(ctx.cascadesContext, ctx.root);
                 }
                 default:
                     throw new IllegalStateException("Table name [" + ctx.root.getTableName() + "] is invalid.");
             }
         }).toRule(RuleType.BINDING_RELATION);
     }
 
-    private LogicalPlan bindWithCurrentDb(CascadesContext cascadesContext, String tableName) {
+    private LogicalPlan bindWithCurrentDb(CascadesContext cascadesContext, UnboundRelation r) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java:72]:
```suggestion
    private LogicalPlan bindWithCurrentDb(CascadesContext cascadesContext, UnboundRelation unboundRelation) {
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java [hunk 3]
```diff
@@ -80,25 +86,39 @@ private LogicalPlan bindWithCurrentDb(CascadesContext cascadesContext, String ta
...
+            }
         } else if (table.getType() == TableType.VIEW) {
             Plan viewPlan = parseAndAnalyzeView(table.getDdlSql(), cascadesContext);
             return new LogicalSubQueryAlias<>(table.getName(), viewPlan);
         }
         throw new AnalysisException("Unsupported tableType:" + table.getType());
     }
 
-    private LogicalPlan bindWithDbNameFromNamePart(CascadesContext cascadesContext, List<String> nameParts) {
+    private LogicalPlan bindWithDbNameFromNamePart(CascadesContext cascadesContext, UnboundRelation r) {
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/BindRelation.java:105]: ditto

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 4]
```diff
@@ -126,6 +135,7 @@ public LogicalOlapScan(RelationId id, Table table, List<String> qualifier,
         this.selectedIndexId = selectedIndexId <= 0 ? getTable().getBaseIndexId() : selectedIndexId;
         this.indexSelected = indexSelected;
         this.preAggStatus = preAggStatus;
+        this.manuallySpecifiedPartitions = partitions;
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:138]:
```suggestion
        this.manuallySpecifiedPartitions = ImmutableList.copyOf(partitions);
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java [hunk 5]
```diff
@@ -101,13 +103,19 @@ public LogicalOlapScan(RelationId id, OlapTable table, List<String> qualifier) {
...
+        this(id, table, qualifier, Optional.empty(), Optional.empty(),
+                specifiedPartitions, false, ImmutableList.of(), false,
+                -1, false, PreAggStatus.on(), specifiedPartitions);
     }
 
     public LogicalOlapScan(RelationId id, Table table, List<String> qualifier) {
         this(id, table, qualifier, Optional.empty(), Optional.empty(),
                 ((OlapTable) table).getPartitionIds(), false, ImmutableList.of(), false,
-                -1, false, PreAggStatus.on());
+                -1, false, PreAggStatus.on(), Collections.emptyList());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/trees/plans/logical/LogicalOlapScan.java:118]:
```suggestion
                -1, false, PreAggStatus.on(), ImmutableList.of());
```

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java [hunk 6]
```diff
@@ -74,11 +72,6 @@ public List<Rule> buildRules() {
                                     IntegerLikeLiteral i = (IntegerLikeLiteral) groupByExpr;
                                     int ord = i.getIntValue();
                                     Expression aggExpr = aggOutput.get(ord - 1);
-                                    if (!CollectionUtils.isEmpty(aggExpr.children())
-                                            && aggExpr.child(0) instanceof Literal) {
-                                        continue;
-                                    }
-                                    checkOrd(ord, aggOutput.size());
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/analysis/ResolveOrdinalInOrderByAndGroupBy.java:81]: why remove these codes in this PR?

File: fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateGroupByConstant.java [hunk 7]
```diff
@@ -50,12 +50,18 @@ public Rule build() {
...
                 expression = FoldConstantRule.INSTANCE.rewrite(expression);
                 if (!(expression instanceof Literal)) {
                     slotGroupByExprs.add(expression);
+                } else {
+                    lit = expression;
                 }
             }
+            if (slotGroupByExprs.isEmpty() && lit != null) {
+                slotGroupByExprs.add(lit);
+            }
```
Comment (morrySnow) [inline fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/logical/EliminateGroupByConstant.java:64]: ditto

**Review by morrySnow — APPROVED** [review]

**Pull request merged**